jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -301,38 +305,50 @@ private Runnable reservedWait()
                 LOG.debug("{} waiting {}", this, ReservedThreadExecutor.this);
 
-            // Keep waiting until stopped, tasked or idle
-            while (_count.getLo() >= 0)
+            // This is now a reserved thread.
+            // Note that this thread must be added to the reserved set
+            // before checking for lo<0 (i.e. stopped), see doStop().
+            _threads.add(this);
+            try
             {
-                try
+                // Keep waiting until stopped, tasked or idle.
+                while (_count.getLo() >= 0)
                 {
-                    // Always poll at some period as safety to ensure we don't poll forever.
-                    Runnable task = _queue.poll(_idleTimeNanos, NANOSECONDS);
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} task={} {}", this, task, ReservedThreadExecutor.this);
-                    if (task != null)
-                        return task;
-
-                    // we have idled out
-                    int size = _count.getLo();
-                    // decrement size if we have not also been stopped.
-                    while (size > 0)
+                    try
                     {
-                        if (_count.compareAndSetLo(size, --size))
-                            break;
-                        size = _count.getLo();
-                    }
-                    _state = size >= 0 ? State.IDLE : State.STOPPED;
-                    return STOP;
+                        // Always poll at some period as safety to ensure we don't poll forever.
+                        Runnable task = _queue.poll(_idleTimeNanos, NANOSECONDS);
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("{} task={} {}", this, task, ReservedThreadExecutor.this);
+                        if (task != null)
+                            return task;
+
+                        // we have idled out
+                        int size = _count.getLo();
+                        // decrement size if we have not also been stopped.
+                        while (size > 0)
+                        {
+                            if (_count.compareAndSetLo(size, --size))
+                                break;
+                            size = _count.getLo();
+                        }
+                        _state = size >= 0 ? State.IDLE : State.STOPPED;
+                        return STOP;
 
+                    }
+                    catch (InterruptedException e)
+                    {
+                        LOG.ignore(e);
+                    }
                 }
-                catch (InterruptedException e)
-                {
-                    LOG.ignore(e);
-                }
+                _state = State.STOPPED;
+                return STOP;
+            }
+            finally
+            {
+                // No longer a reserved thread.
+                _threads.remove(this);
```

Comment:
I don't like this extra work here.  We want minimum time and effort between allocating a reserved thread and it acting on the task.
Making it modifyf the `_threads` collection here just wastes time.  See my alternative cleanup of dump in #6658